elasticdl/python/common/save_utils.py

```diff
@@ -0,0 +1,252 @@
+            return None
+
+        version_folders = os.listdir(checkpoint_dir)
+        if not version_folders:
+            return None
+        version_num = [int(v.split("-")[-1]) for v in version_folders]
+        version_folder_pairs = sorted(
+            zip(version_num, version_folders), reverse=True
+        )
+        for version, folder in version_folder_pairs:
+            folder_dir = os.path.join(checkpoint_dir, folder)
+            if CheckpointSaver.check_checkpoint_valid(folder_dir):
+                return folder_dir
+        return None
+
+    @staticmethod
+    def check_checkpoint_valid(checkpoint_dir):
+        """Check whether the checkpoint directory is valid. The filename template
+        in the checkpoint directory like "variables-{i}-of-{N}.ckpt". We will
+        parse any filename to get N which is the total number of parameters
+        shards. It is valid if the number of files in the directory N.
+        """
+        if not os.path.exists(checkpoint_dir):
+            return False
+
+        shard_files = os.listdir(checkpoint_dir)
+        if not shard_files:
+            return False
+
+        shard_file_prefix = shard_files[0].split(".")[0]
+        expected_shard_num = int(shard_file_prefix.split("-")[-1])
+        return expected_shard_num == len(shard_files)
+
+    @staticmethod
+    def restore_params_from_checkpoint(checkpoint_dir, shard_index, shard_num):
+        """Restore a shard parameters from the checkpoint directory.
+        If shard_num=1, a entire model parameters will be restored.
+
+        Args:
+            checkpoint_dir: a directory with checkpoint files.
+            shard_index: Model shard index, e.g. the PS instance index
+                using ParameterServerStrategy with multiple PS instances.
+            shard_num: The total number of model shards, e.g. the total PS
+                instancecount using ParameterServerStrategy with multiple
+                PS instances.
+
+        Return:
+            parameters: A Parameter object which contains model version,
+                non-embedding parameters and embedding tables for the
+                PS instance with ps_id.
+        """
+        from elasticdl.python.ps.embedding_table import create_embedding_table
+
+        variable_shard_files = os.listdir(checkpoint_dir)
+        non_embedding_vars = {}
+        embedding_tables = {}
+        version = None
+        for shard_file in variable_shard_files:
+            shard_file_path = os.path.join(checkpoint_dir, shard_file)
+            model_pb = elasticdl_pb2.Model()
+            model_pb = load_pb_from_file(model_pb, shard_file_path)
+            if version is None:
+                version = model_pb.version
+            elif version != model_pb.version:
+                raise ValueError(
+                    "The versions in model shards are not consistency"
+                )
+
+        for embedding_info_pb in model_pb.embedding_table_info:
```

Comment:
Put this in the first for loop?

```python
for shard_file in variable_shard_files:
    ....
    for embedding_info_pb in model_pb.embedding_table_info:
        ....
```

Reply:
Yes, Thanks! It is a mistake when I move the code.